src/features/generic_events/aggregate/index.js

```diff
@@ -0,0 +1,114 @@
+import { getConfigurationValue, getInfo, getRuntime } from '../../../common/config/config'
+import { FEATURE_NAME } from '../constants'
+import { isBrowserScope } from '../../../common/constants/runtime'
+import { AggregateBase } from '../../utils/aggregate-base'
+import { warn } from '../../../common/util/console'
+import { now } from '../../../common/timing/now'
+import { deregisterDrain } from '../../../common/drain/drain'
+
+export class Aggregate extends AggregateBase {
+  #agentRuntime
+  static featureName = FEATURE_NAME
+  constructor (agentIdentifier, aggregator) {
+    super(agentIdentifier, aggregator, FEATURE_NAME)
+
+    this.eventsPerHarvest = 1000
+    this.harvestTimeSeconds = getConfigurationValue(this.agentIdentifier, 'generic_events.harvestTimeSeconds')
+
+    this.referrerUrl = undefined
+    this.currentEvents = []
+
+    this.events = []
+    this.overflow = []
+
+    this.#agentRuntime = getRuntime(this.agentIdentifier)
+
+    if (isBrowserScope && document.referrer) this.referrerUrl = cleanURL(document.referrer)
+
+    this.waitForFlags(['ins']).then(([ins]) => {
+      if (!ins) {
+        this.blocked = true
+        deregisterDrain(this.agentIdentifier, this.featureName)
+        return
+      }
+
+      // handle page actions and other generic events here
+      this.harvestScheduler = new HarvestScheduler('ins', { onFinished: (...args) => this.onHarvestFinished(...args) }, this)
+      this.harvestScheduler.harvest.on('ins', (...args) => this.onHarvestStarted(...args))
+      // this.harvestScheduler.startTimer(this.harvestTimeSeconds, 0)
+
+      this.drain()
+    })
+  }
+
+  onHarvestStarted (options) {
+    const { userAttributes, atts } = getInfo(this.agentIdentifier)
+    const harvestEvents = this.overflow.length ? this.overflow.splice(0, Infinity) : this.events.splice(0, Infinity)
+    var payload = ({
+      qs: {
+        ua: userAttributes,
+        at: atts
+      },
+      body: {
+        ins: harvestEvents
+      }
+    })
+
+    if (options.retry) {
+      this.currentEvents = harvestEvents
+    }
+
+    return payload
+  }
+
+  onHarvestFinished (result) {
+    if (result && result.sent && result.retry && this.currentEvents.length) {
+      this.events = this.currentEvents.concat(this.events)
+      this.currentEvents = []
+    }
+  }
+
+  // WARNING: Insights times are in seconds. EXCEPT timestamp, which is in ms.
+  addEvent (obj = {}) {
+    if (!obj || !Object.keys(obj).length) return
+    if (!obj.eventType) {
+      warn('Invalid object passed to generic event aggregate. Missing "eventType".')
+      return
+    }
+
+    for (let key in obj) {
+      let val = obj[key]
+      if (key === 'timestamp') val = this.#agentRuntime.timeKeeper.correctAbsoluteTimestamp(val)
+      obj[key] = (val && typeof val === 'object' ? stringify(val) : val)
+    }
+
+    const eventAttributes = {
+      /** Agent-level custom attributes */
+      ...(getInfo(this.agentIdentifier).jsAttributes || {}),
```

Comment:
Should jsAttributes be last in the list so that it gets truncated if the size is too large?

Reply:
yes and also so that it matches the same attribute precedence